File: src/zojax/formatter/resources/humandatetime.js
var formatter_past_dictionary = {
    '%s year(s) ago'  : '%s year(s) ago',
    '%s month(s) ago'  : '%s month(s) ago',
    '%s week(s) ago': '%s week(s) ago',
    '%s day(s) ago' : '%s day(s) ago',
    '%s hour(s) ago' : '%s hour(s) ago',
    '%s minute(s) ago' : '%s minute(s) ago',
    '%s second(s) ago' : '%s second(s) ago'
};

var formatter_future_dictionary = {
    '%s year(s) ago'  : '%s year(s) ago',
    '%s month(s) ago'  : '%s month(s) ago',
    '%s week(s) ago': '%s week(s) ago',
    '%s day(s) ago' : '%s day(s) ago',
    '%s hour(s) ago' : '%s hour(s) ago',
    '%s minute(s) ago' : '%s minute(s) ago',
    '%s second(s) ago' : '%s second(s) ago'
};

function setHumandatetimeFormatter(el)
{
    var time = new Date();
    time.setTime(Date.parse(el.attr('value')));
    var new_time = new Date();
    delta = new_time-time;
    $.i18n.setDictionary(formatter_past_dictionary);
    if (delta < 0) {
        delta = Math.abs(delta);
        $.i18n.setDictionary(formatter_future_dictionary)
    }
    years = Math.floor(delta/(365*24*60*60*1000.0));
    months = Math.floor(delta/(30*24*60*60*1000.0));
    weeks = Math.floor(delta/(7*24*60*60*1000.0));
    days = Math.floor(delta/(24*60*60*1000.0));
    hours = Math.floor(delta/(60*60*1000.0));
    minutes = Math.floor(delta/(60*1000.0));
    seconds = Math.floor(delta/(1000.0));
    var res;
    if (years)
        res = $.i18n._('%s year(s) ago', [years])
    else if (months)
        res = $.i18n._('%s month(s) ago', [months])
    else if (weeks)
        res = $.i18n._('%s week(s) ago', [weeks])
    else if (days)
        res = $.i18n._('%s day(s) ago', [days])
    else if (hours)
        res = $.i18n._('%s hour(s) ago', [hours])
    else if (minutes)
        res = $.i18n._('%s minute(s) ago', [minutes])
    else
        res = $.i18n._('%s second(s) ago', [seconds])
    el.text(res);
    el.attr('processed', 'true')
}


$(document).ready(function() {
    var elems = $("span.zojax-formatter-humandatetime");
    for (var i = 0; i < elems.length; i++)
    {
        var el = $(elems[i]);
        if (!el.attr('processed')) {
            setHumandatetimeFormatter(el);
        }
    }
});

File: src/zojax/formatter/humandatetime.py
##############################################################################
#
# Copyright (c) 2009 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
""" ``humanDatetime`` formatter implementation

$Id$
"""
from pytz import utc, timezone, FixedOffset
from datetime import datetime, timedelta
from zope import interface, component
from zope.i18n import translate
from zope.component import getUtility
from zope.publisher.interfaces.http import IHTTPRequest

from interfaces import _, IFormatter, IFormatterFactory, IFormatterConfiglet


class HumanDatetimeFormatter(object):
    interface.implements(IFormatter)

    messages = {'past': {'year': u'${value} year(s) ago',
                         'month': u'${value} month(s) ago',
                         'week': u'${value} week(s) ago',
                         'day': u'${value} day(s) ago',
                         'hour': u'${value} hour(s) ago',
                         'minute': u'${value} minute(s) ago',
                         'second': u'${value} second(s) ago',
                         },
                'future': {'year': u'in ${value} year(s)',
                           'month': u'in ${value} month(s)',
                           'week': u'in ${value} week(s)',
                           'day': u'in ${value} day(s)',
                           'hour': u'in ${value} hour(s)',
                           'minute': u'in ${value} minute(s)',
                           'second': u'in ${value} second(s)',
                           }}

    def __init__(self, request, *args):
        self.request = request

    def format(self, value):
        configlet = getUtility(IFormatterConfiglet)

        tz = timezone(configlet.timezone)

        if value.tzinfo is None:
            value = utc.localize(value)

        value = value.astimezone(tz)

        now = datetime.now(utc)
        delta = now - value.astimezone(utc)
        key = 'past'
        if delta < timedelta():
            delta = - delta + timedelta(seconds=1) #due to python implementation
            key = 'future'

        offset = value.tzinfo.utcoffset(value)
        if offset < timedelta():
            ind = -1
        else:
            ind = 1
        offset = ind*(abs(offset).seconds/600)*10

        value = FixedOffset(offset).normalize(value)

        years, months, weeks, hours, minutes = (
            delta.days/365, delta.days/30, delta.days/7,
            delta.seconds/3600, delta.seconds/60)
        formatted = None

        if years > 0:
            formatted = translate(self.messages[key]['year'], 'zojax.formatter',
                                  mapping={'value': years})
        elif months > 0:
            formatted = translate(self.messages[key]['month'],'zojax.formatter',
                                  mapping={'value': months})
        elif weeks > 0:
            formatted = translate(self.messages[key]['week'], 'zojax.formatter',
                                  mapping={'value': weeks})
        elif delta.days > 0:
            formatted = translate(self.messages[key]['day'], 'zojax.formatter',
                                  mapping={'value': delta.days})
        elif hours > 0:
            formatted = translate(self.messages[key]['hour'], 'zojax.formatter',
                                  mapping={'value': hours})
        elif minutes > 0:
            formatted = translate(
                self.messages[key]['minute'], 'zojax.formatter',
                mapping={'value': minutes})
        else:
            formatted = translate(
                self.messages[key]['second'], 'zojax.formatter',
                mapping={'value': delta.seconds})

        return """<span class="zojax-formatter-humandatetime" value="%s">%s</span>""" \
                % (value.strftime('%B %d, %Y %H:%M:%S %z'), formatted)


class HumanDatetimeFormatterFactory(object):
    component.adapts(IHTTPRequest)
    interface.implements(IFormatterFactory)

    def __init__(self, request):
        self.request = request

    def __call__(self, *args, **kw):
        return HumanDatetimeFormatter(self.request, *args)
